# Funciones

En el contexto de la programación, una *función* es una secuencia de declaraciones que realiza un cálculo y esa secuencia tiene nombre. Cuando se define una función, se especifica el nombre y la secuencia de declaraciones. Más tarde, se puede "llamar" a la función por su nombre.



## 1. Llamadas a funciones

Ya hemos visto un ejemplo de llamada a función. Si ejecutamos
```
type(42)
```
se imprime
```
<clase 'int'>
```
El nombre de la función es `type`. La expresión entre paréntesis
se llama el *argumento* de la función. El resultado, para esta función, es el tipo del argumento.

Es común decir que una función "toma" un argumento y "devuelve" un resultado. El resultado también se denomina *valor de retorno*. 

En  el caso de `type` es una función que toma cualquier valor y  devuelve el tipo del valor. 

Otra función que hemos visto es `print`. Esta función toma un  argumento  o varios separados por comas e imprime en pantalla su valor o, en el caso de varios  argumentos,  sus valores separados por espacios. Es una función que no devuelve nada,  es decir no devuelve un valor que pueda ser utilizado posteriormente. Profundizaremos el concepto de "funciones que no devuelven nada" más adelante. 

Ejemplos de la función `print`:

In [1]:
print('23')
x = -1 
print(x)
print('hola','adiós')

23
-1
hola adiós


Python proporciona funciones que convierten valores de un tipo a otro. La función `int` toma cualquier valor y si puede lo convierte en un número entero. En caso que no pueda se produce un error. Todo esto se puede verificar ejecutando la siguiente celda de código:

In [4]:
print(int('32'))
# int('Hola') # descomentar la línea produce un error

32



`int` puede convertir valores de coma flotante en números enteros, pero no redondea; corta la parte decimal:


In [5]:
print(int(3.99999))
print(int(-2.3))

3
-2


`float` convierte enteros y cadenas que representan números en números punto flotante:


In [6]:
print(float(32))
print(float('3.14159'))
# print(float('hola')) # descomentar la línea produce error

32.0
3.14159


Finalmente, `str` convierte cualquier argumento válido en una cadena:

In [7]:
print(str(32), type(str(32)))
print(str(3.14159), type(str(3.14159)))

32 <class 'str'>
3.14159 <class 'str'>


## 2. Funciones matemáticas

Python tiene un módulo matemático que proporciona la mayoría de las funciones matemáticas familiares. Un *módulo* o *biblioteca* es un archivo que contiene una colección de funciones relacionadas. 

Antes de que podamos usar las funciones en un módulo, tenemos que importarlo con una declaración de importación. Para poder usar el módulo `math` en este cuaderno Jupyter debemos ejecutar la siguiente celda de código:


In [8]:
import math

Esta declaración crea un *objeto módulo* llamado math. Si imprime el objeto módulo, obtendrá información sobre él:


In [9]:
print(math)

<module 'math' (built-in)>


El objeto módulo contiene las funciones y variables definidas en el
módulo. Para acceder a una de las funciones, debe especificar el nombre
del módulo y el nombre de la función, separados por un punto (también
conocido como período). Este formato se llama *notación punto* o *dot  notation* en inglés.
```
radio = potencia_señal / potencia_ruido
decibeles = 10 * math.log10 (radio)
```

```
radianes = 0.7
altura = math.sin(radianes)
```
El primer ejemplo usa `math.log10` para calcular una relación señal-ruido en decibelios (asumiendo que `potencia_señal` y `potencia_ruido` están definidos). El módulo math también proporciona `log`, que calcula logaritmos en base `e`.

El segundo ejemplo encuentra el seno de `radianes`. El nombre de la variable `radianes` es una pista de que `sin` (seno) y las otras funciones trigonométricas (`cos`, `tan`, etc.) toman argumentos en radianes. Para convertir de grados a radianes, hay que dividir por 180 y multiplicar por $\pi$. Para ejemplificar,  ejecute el siguiente código:


In [10]:
grados = 45
radianes = grados / 180.0 * math.pi
math.sin(radianes)

0.7071067811865476

La expresión `math.pi` obtiene la variable `pi` del módulo math. Su valor es una aproximación en punto flotante de $\pi$, con una precisión de proximadamente 15 dígitos.

Usando trigonometría, podemos verificar el resultado anterior comparándolo con la raíz cuadrada de dos dividida por dos:


In [11]:
math.sqrt(2) / 2.0

0.7071067811865476

Para convertir de radianes a grados en formato decimal se hace la operación inversa: para convertir de grados a radianes dividíamos por 180  y multiplicábamos por $\pi$. Ahora debemos multiplicar por 180 y dividir por $\pi$. 

In [12]:
radianes = math.pi / 6
grados = 180.0 * radianes / math.pi
print(grados)

29.999999999999996


## 3. Composición

Hasta ahora, hemos examinado los elementos de un programa; variables, expresiones y declaraciones; de forma aislada, sin hablar de cómo combinarlos.

Una de las características más útiles de los lenguajes de programación es su capacidad para tomar pequeños bloques de construcción y *componer*. Por ejemplo, el argumento de una función puede ser cualquier tipo de expresión, incluidos los operadores aritméticos:
```
x = math.sin(grados / 360.0 * 2 * math.pi)
```
E incluso llamadas a funciones:
```
x = math.exp (math.log (x + 1))
```
Casi en cualquier lugar donde se pueda poner un valor, se puede poner una expresión arbitraria, con una excepción: el lado izquierdo de una declaración de asignación tiene que ser un nombre de variable. Cualquier otra expresión en el lado izquierdo es un error de sintaxis (veremos excepciones a esta regla más adelante). Si ejecutamos
```
minutos = horas * 60 # correcto
horas * 60 = minutos # incorrecto!
```
obtenemos 
```
SyntaxError: can't assign to operator
```


## 4. Añadiendo nuevas funciones

Hasta ahora, solo hemos estado usando las funciones incorporadas en Python, pero también es posible agregar nuevas funciones. Una definición de función especifica el nombre de una nueva función y la secuencia de instrucciones que se ejecutan cuando se llama a la función.

Veamos el siguiente  ejemplo:

```
def imprimir_instrucciones():
    print ("Abra el  recipiente.")
    print ("Complete el recipiente con agua.")
```
`def` es una palabra clave que indica que se trata de una definición de función. El nombre de la función es `imprimir_instrucciones`. 

Las reglas para los nombres de las funciones son las mismas que para los nombres de las variables: las letras, los números y el subrayado son legales, pero el primer carácter no puede ser un número. No puede usar una palabra clave como nombre de una función, y debe evitar tener una variable y una función con el mismo nombre.

Los paréntesis vacíos después del nombre indican que esta función no acepta argumentos.

La primera línea de la definición de la función se llama *header* o *encabezado*; el resto se llama *body* o *cuerpo de la función*. El encabezado debe terminar con dos puntos y el cuerpo debe tener sangría. Por convención, la sangría es siempre de cuatro espacios. El cuerpo puede contener cualquier número de declaraciones. 

Hablando propiamente del cuerpo de la función `imprimir_instrucciones()`, las cadenas de las declaraciones de impresión están entre comillas dobles. Las comillas simples y las dobles hacen lo mismo; la mayoría de la gente usa comillas simples. En  el caso  donde aparece una comilla simple (que también es un apóstrofe) en la cadena se utilizan las comillas dobles (este no es el caso).Todas las comillas (simples y dobles) deben ser "comillas rectas" las "comillas inclinadas compuestas de tildes (como ser \` o bien ´) no son legales en Python.


La definición de una función crea un *objeto función*, que tiene el tipo `function`. Pruebe con el siguiente código:


In [2]:
def imprimir_instrucciones():
    print ("Abra el  recipiente.")
    print ("Complete el recipiente con agua.")

print(imprimir_instrucciones)
type(imprimir_instrucciones)

<function imprimir_instrucciones at 0x0000015A41AFC670>


function

La sintaxis para llamar a la nueva función es la misma que para las funciones integradas:


In [14]:
imprimir_instrucciones()

Abra el  recipiente.
Complete el recipiente con agua.


Una vez que haya definido una función, puede usarla dentro de otra función. Por ejemplo, para repetir el texto anterior, podríamos escribir una función llamada `repetir_instrucciones`:

In [15]:
def repetir_instrucciones():
    imprimir_instrucciones()
    imprimir_instrucciones()

Si ejecutamos la celda anterior y la siguiente repetimos las instrucciones:

In [16]:
repetir_instrucciones()

Abra el  recipiente.
Complete el recipiente con agua.
Abra el  recipiente.
Complete el recipiente con agua.


Al reunir los fragmentos de código, todo el programa se ve así:


```
def imprimir_instrucciones():
    print ("Abra el  recipiente.")
    print ("Complete el recipiente con agua.")

def repetir_instrucciones():
    imprimir_instrucciones()
    imprimir_instrucciones()

repetir_instrucciones()
```
Este programa contiene dos definiciones de función: `imprimir_instrucciones` y `repetir_instrucciones`. Las definiciones de funciones se ejecutan al igual que otras declaraciones, pero el efecto es crear objetos de función. Las declaraciones dentro de la función no se ejecutan hasta que se llama a la función y la definición de la función no genera salida.

Como era de esperar, debe crear una función antes de poder ejecutarla. En otras palabras, la definición de la función debe ejecutarse antes de que se llame a la función.

Como ejercicio, mueva la última línea de este programa hacia arriba, para que la llamada a la función aparezca antes de las definiciones. Ejecute el programa y vea qué mensaje de error obtiene.

Ahora mueva la llamada a la función al final y mueva la definición de `imprimir_instrucciones` después de la definición de `repetir_instrucciones`. ¿Qué sucede cuando ejecuta este programa?




## 5. Cadenas de documentación en Python

Una *cadena de documentación* es un tipo de comentario que comienza y termina  con `"""`. Por  ejemplo

    """Comentario en una línea"""
   
o

    """
    Esta es una línea de comentario
    Esta también
    ...
    """

Esta forma de comentar permite hacer comentarios extensos y además lo que se encuentra en estos comentarios permite generar automáticamente la documentación del código, tema que no veremos en este curso.

Las cadenas de documentación se llaman *docstrings*.

Técnicamente hablando, los docstrings no son comentarios. Crean variables anónimas que hacen referencia a las cadenas. Además, no son ignoradas por el intérprete de Python. Como dijimos más arriba se utilizan para generar documentación.

Python proporciona dos tipos de docstrings: docstrings de una línea y docstrings de varias líneas.

### 1) Docstrings de una línea

Como su nombre indica, un docstring de una línea cabe en una línea. Un docstring de una línea comienza con comillas triples (""") y también termina con comillas triples ("""). Además, no habrá ninguna línea en blanco ni antes ni después de la docstring de una línea.

El siguiente ejemplo ilustra un docstring de una línea en la función quicksort():

    def quicksort():
        """ ordenar la lista usando el algoritmo quicksort """
        ...

### 2) Docstrings multilínea

A diferencia de los docstrings de una línea, los docstrings de varias líneas pueden abarcar varias líneas. Una docstring multilínea también comienza con comillas triples (`"""`) y termina con comillas triples (`"""`).

El siguiente ejemplo muestra cómo utilizar las cadenas de documentos multilínea:

    def incrementar(sueldo, rating):
        """ 
        pre: sueldo - número decimal
             rating - número entero de 1 al 6
        post: incrementa el sueldo en base rating
                rating 1 - 2 no incrementa
                rating 3 - 4 incrementa 5%
                rating 4 - 6 incrementa 10%
        """

## 5. Ejemplo: conversión de formatos de grados

Las coordenadas de los mapas se representaban tradicionalmente como grados, minutos y segundos (GMS), llamados  *grado sexagesimales*. Sin embargo, en el SIG (que está basado en el ordenador), la latitud y la longitud se representan como números decimales conocidos como *grados decimales*. El formato de grados, minutos y segundos se sigue utilizando. A veces, hay que convertir entre ese formato y los grados decimales para realizar cálculos y elaborar informes.

Escribiremos las funciones de conversión entre ambos sistemas. 

Primero convertiremos grados sexagesimales en  grados decimales y eso es muy sencillo cuando usamos grados positivos:

In [17]:
def sexa2deci(grados: int , minutos: int , segundos: float) -> float:
    # pre: 0 <= grados, 0 <= minutos < 60, 0 <= segundos < 60
    # post: devuelve los grados en notación decimal
    return grados + minutos / 60 + segundos / 3600 

print(sexa2deci(33, 54, 23.5))

33.906527777777775


Un poco más laboriosa es la conversión de grados decimales a sexagesimales:

In [18]:
import math 

def deci2sexa(pos: float) -> tuple:
    # pre: pos son grados en notación decimal, pos >= 0
    # post:  devuelve pos en una 3-upla grados, minutos, segundos
    grados = math.floor(pos)
    resto = pos - grados
    minutos = math.floor(resto * 60)
    resto = resto * 60 - minutos
    segundos = 60 * resto
    return grados, minutos, segundos

print(deci2sexa(33.906527777777775))

(33, 54, 23.499999999990564)


La función `math.floor()` trunca un número positivo a su parte entera. 

Las dos funciones anteriores deben considerarse funciones auxiliares que ayudarán a resolver el problema, un poco más complejo, de pasar de coordenadas terrestres sexagesimales a coordenadas terrestres decimales y viceversa. Eso lo podremos hacer usando condicionales, los cuales veremos dentro de poco.

## 6. Flujo de ejecución

Para asegurarse de que una función esté definida antes de su primer uso, debe conocer las instrucciones de orden en que se ejecutan, lo que se denomina flujo de ejecución.

La ejecución siempre comienza en la primera declaración del programa. Las declaraciones se ejecutan de una en una, en orden de arriba hacia abajo.

Las definiciones de funciones no alteran el flujo de ejecución del programa, pero hay que recordar que las declaraciones dentro de la función no se ejecutan hasta que se llama a la función.

Una llamada a una función es como un desvío en el flujo de ejecución. En lugar de ir a la siguiente declaración, el flujo salta al cuerpo de la función, ejecuta las declaraciones allí y luego regresa para continuar donde lo dejó.

Eso suena bastante simple, hasta que recuerda que una función puede llamar a otra. Mientras esté en medio de una función, el programa podría tener que ejecutar las instrucciones en otra función. Luego, mientras se ejecuta esa nueva función, ¡es posible que el programa tenga que ejecutar otra función más!

Afortunadamente, Python es bueno para realizar un seguimiento de dónde se encuentra, por lo que cada vez que se completa una función, el programa retoma donde lo dejó en la función que la llamó. Cuando llega al final del programa, termina.

En resumen, cuando lees un programa, no siempre hay leer de arriba a abajo. En  general tiene más sentido seguir el flujo de ejecución.

## 7. Parámetros y argumentos

Algunas de las funciones que hemos visto requieren argumentos. Por ejemplo, cuando llamas a `math.sin` pasás un número como argumento. Algunas funciones toman más de un argumento: `math.pow` toma dos, la base y el exponente.

Dentro de la función, los argumentos se asignan a variables llamadas *parametros*. Aquí hay una definición de una función que toma un argumento:



In [19]:
# Ejecutar esta función
def imprimir_4_veces(nombre):
    print(nombre)
    print(nombre)
    print(nombre)
    print(nombre)

Esta función asigna el argumento a un parámetro llamado `nombre`. Cuando se llama a la función, imprime el valor del parámetro (cualquiera que sea) cuatro veces.

La función trabaja con cualquier valor que se pueda imprimir. Por ejemplo:


In [20]:
imprimir_4_veces('Spam')
imprimir_4_veces(42)
imprimir_4_veces(math.pi)

Spam
Spam
Spam
Spam
42
42
42
42
3.141592653589793
3.141592653589793
3.141592653589793
3.141592653589793


Las mismas reglas de composición que se aplican a las funciones predeefinidas también se aplican a las funciones definidas por el programador, por lo que podemos usar cualquier tipo de expresión como argumento para `imprimir_4_veces`:

In [21]:
imprimir_4_veces('Spam' * 2)
imprimir_4_veces(42 + 3)
imprimir_4_veces(math.cos(math.pi))

SpamSpam
SpamSpam
SpamSpam
SpamSpam
45
45
45
45
-1.0
-1.0
-1.0
-1.0


El argumento se evalúa antes de llamar a la función, por lo que en los ejemplos las expresiones `'Spam' * 2`, `42 + 3`y `math.cos(math.pi)` solo se evalúan una vez.

También se puede utilizar una variable como argumento:

In [22]:
miguel = 'el hombre que escaló la montaña'
imprimir_4_veces(miguel)

el hombre que escaló la montaña
el hombre que escaló la montaña
el hombre que escaló la montaña
el hombre que escaló la montaña


El nombre de la variable que pasamos como argumento (`miguel`) no tiene nada que ver con el nombre del parámetro (`nombre`). No importa cuál sea el nombre del argumento que se utilizó en la definición; cuando aplicamos la función `imprimir_4_veces`  al parámetro `miguel` toda ocurrencia de `nombre` es cambiada por `miguel`.

Para ejecuciones repetidas  de una misma instrucción o similares, podemos hacer un código más corto y mas legible utilizando la instrucción `for`: 


In [23]:
def imprimir_4_veces_v2(nombre):
    for _ in range(4):
        print(nombre)

Si ejecutamos


In [24]:
imprimir_4_veces_v2('¡Hola!')

¡Hola!
¡Hola!
¡Hola!
¡Hola!


debería ver algo como esto:
```
¡Hola!
¡Hola!
¡Hola!
¡Hola!
```

Este es el uso más simple de la instrucción `for` que veremos con más profundidad después. Pero esto que explicamos es suficiente para permitirnos escribir un programa que generaliza la función anterior a un número determinado de repeticiones. Es decir,  podemos definir la función
```
imprimir_veces(n, nombre)
```
que imprime `n` veces la cadena nombre:

In [25]:
def imprimir_veces(n, nombre):
    for _ in range(n):
        print(nombre)

imprimir_veces(5, '¡Adios!')

¡Adios!
¡Adios!
¡Adios!
¡Adios!
¡Adios!


La sintaxis de una instrucción `for` es similar a la definición de una función. Tiene un encabezado que termina con dos puntos y un cuerpo justificado, generalmente a 4 espacios del `for`. El cuerpo puede contener cualquier número de declaraciones.

Una instrucción `for` también se llama un *ciclo* porque el flujo de ejecución corre a través del cuerpo y luego regresa a la parte superior. En el caso de `imprimir_4_veces_v2` recorre el cuerpo cuatro veces. En  el caso de `imprimir_veces(n, nombre)` recorre el cuerpo la cantidad de veces que especifiquemos que vale `n`.


## 8. Ejemplo: medición de distancias en la Tierra


La esencia del análisis geoespacial es descubrir las relaciones de los objetos en la Tierra. Los objetos que están más cerca tienden a tener una relación más fuerte que los que están más lejos más alejados. Este concepto se conoce como la Primera Ley de la Geografía de Tobler. Por lo tanto, la medición de la distancia es una función crítica del análisis geoespacial.

Como hemos aprendido, cada mapa es un modelo de la Tierra y todos están equivocados en cierta medida. Por eso, medir la distancia exacta entre dos puntos de la Tierra
sentado frente a un ordenador es imposible. Incluso los topógrafos profesionales (que salen al campo tanto con equipos de observación tradicionales como con equipos de GPS muy precisos) no tienen en cuenta todas las anomalías de la superficie terrestre entre el punto A y el punto B. Por tanto, para medir la distancia, debemos plantearnos las siguientes preguntas:

- ¿Qué estamos midiendo?
- ¿Cuánto estamos midiendo?
- ¿Qué precisión necesitamos?

Ahora bien, para calcular la distancia, hay tres modelos de la Tierra que podemos utilizar:

- Plano
- Esférico
- Elipsoide

En el modelo plano, se utiliza la geometría euclidiana estándar. La Tierra se considera un plano sin curvatura, como se muestra en el siguiente diagrama:

![](../geospatial_ana/figures/01_world.png)

Este modelo hace que las matemáticas sean bastante sencillas porque se trabaja con líneas rectas. El formato más común para las coordenadas geoespaciales es el de los grados decimales. Sin embargo, las coordenadas de grados decimales son medidas de referencia en una esfera tomadas como ángulos -entre la longitud y el primer meridiano- y la latitud y el ecuador. Además, las líneas de longitud convergen hacia el cero en los polos. La circunferencia de cada línea de latitud también se reduce hacia los polos. Estos hechos significan que los grados decimales no son un sistema de coordenadas válido sistema de coordenadas válido para la geometría euclidiana, que utiliza el plano.

Las proyecciones cartográficas intentan simplificar los problemas de tratar con un elipsoide 3D en un plano 2D, ya sea en papel o en una pantalla de ordenador. Las proyecciones cartográficas aplanan un modelo redondo de la Tierra a un plano e introducen distorsiones a cambio de la comodidad de contar con un mapa en una hoja. Una vez que esta proyección está en su lugar y los grados decimales se cambian por un sistema de coordenadas cartesianas con coordenadas $x$ e $y$, podemos utilizar las formas más simples de la geometría euclidiana, es decir, el teorema de Pitágoras. 

A una escala suficientemente grande, una esfera o elipsoide como la Tierra se parece más a un plano que a una esfera. De hecho, durante siglos, todo el mundo pensó que la Tierra era plana. Si la diferencia en grados de longitud es lo suficientemente pequeña, a menudo se puede utilizar la geometría euclidiana y luego convertir las medidas en metros, kilómetros o millas. Este método generalmente no se recomienda, pero la decisión depende en última instancia de usted y de sus requisitos de precisión como analista.

El enfoque del modelo esférico trata de aproximarse mejor a la realidad evitando los problemas derivados de aplastar la Tierra en una superficie plana. Como su nombre indica, este modelo utiliza una esfera perfecta para representar la Tierra (similar a un globo terráqueo físico), lo que nos permite trabajar con grados directamente. Este modelo ignora el hecho de que la Tierra es en realidad más bien un elipsoide con diferentes grados de grosor en su corteza. Pero al trabajar con la distancia en la superficie de una esfera, podemos empezar a medir distancias más largas con más precisión. La siguiente captura de pantalla ilustra este concepto:

![](../geospatial_ana/figures/02_esfera.png)

Utilizando el modelo de elipsoide de la Tierra, los analistas se esfuerzan por conseguir el mejor modelo de la superficie terrestre. Existen varios modelos de elipsoide, que se denominan datums. Un *datum* es un conjunto de valores que definen una forma estimada de la Tierra, también conocido como *sistema geodésico*. Como cualquier otro sistema de georreferenciación, un datum puede ser optimizado para un área localizada. El datum más utilizado es el llamado *[WGS84](https://es.wikipedia.org/wiki/WGS84)*, que está diseñado para un uso global. Debe saber que el WGS84 se actualiza ocasionalmente a medida que mejoran las técnicas de evaluación y la tecnología. La revisión más reciente se produjo en 2004.

En América del Norte, se utiliza el datum NAD83 para optimizar la referenciación en el continente. En el hemisferio oriental, se utiliza con más frecuencia el *Sistema de Referencia Terrestre Europeo de 1989 (ETRS89)*. El ETRS89 está fijado a la parte estable de la placa euroasiática. Los mapas de Europa basados en el ETRS89 son inmunes a la deriva continental, que cambia hasta 2,5 cm por año a medida que la corteza terrestre se desplaza.

Un elipsoide no tiene un radio constante desde el centro. Este hecho significa que las fórmulas utilizadas en el modelo esférico de la Tierra empiezan a tener problemas en el modelo elipsoidal. Aunque no es una aproximación perfecta, se acerca mucho más a la realidad que el modelo esférico. La siguiente captura de pantalla muestra un modelo genérico de elipsoide denotado por una línea negra contrastada con una representación de la corteza terrestre irregular, que utiliza una línea roja para representar el geoide. Aunque no lo utilizaremos para estos ejemplos, otro modelo es el del geoide. El geoide es el modelo más preciso y exacto de la Tierra, que se basa en la superficie terrestre sin ningún factor de influencia, excepto la gravedad y la rotación. El siguiente diagrama es una representación de un modelo geoide, elipsoide y esférico para ilustrar sus diferencias:

![](../geospatial_ana/figures/03_geoide.png)

Entender estos modelos de la Tierra es fundamental porque, después de todo, estamos modelando la Tierra. 

Ahora que hemos discutido estos diferentes modelos de la Tierra y los problemas para medirlos, veamos algunas soluciones usando Python.


### La fórmula del haverseno


Aunque la modelización matemática más exacta de la Tierra es como un elipsoide de revolución, para nuestros fines nos resultará útil, y más sencillo,  pensar a la Tierra como una esfera de radio 6.371 km. 

La distancia entre dos puntos en la Tierra viene dada por la longitud de la porción del *círculo máximo*, también llamado *geodésica*,  que los une. Los meridianos son los círculos máximos que pasan por los  polos,  pero hay otros círculos máximos. El ecuador terrestre es uno. En general dados dos puntos cualesquiera de la Tierra hay un círculo máximo que los contiene: es la intersección de la esfera con  el plano determinado por los dos puntos y el centro de la esfera.  Las círculos máximos en la esfera juegan el rol de las rectas en el plano y ellas permiten definir triángulos esféricos, cuadrados esféricos, etc. y,  en la denominada geometría esférica,  se pueden demostrar propiedades análogas a las que se obtienen en la geometría del plano.  Un *arco geodésico* entre dos puntos es la porción de círculo máximo que une los dos puntos y tiene longitud menor. La *distancia* en la Tierra entre dos puntos es la longitud del arco geodésico que une los dos puntos.

Entonces ¿cómo podemoes estimar medir la distancia entre dos puntos? El método más popular es utilizar la *fórmula del haverseno* o *semiverseno*, que utiliza la trigonometría para calcular la longitud del arco geodésico entre dos puntos utilizando como entrada coordenadas definidas en grados decimales. 

Utilizaremos el módulo `math`que nos permite utilizar las funciones seno y coseno,  necesarias para implementar la fórmula del haverseno.

In [26]:
from math import * # importa la biblioteca o módulo de matemática
def distancia_h(lon1, lat1, lon2, lat2):
    x_dist = (pi / 180) * (lon1 - lon2)
    y_dist = (pi / 180) * (lat1 - lat2)
    y1_rad = (pi / 180) * (lat1)
    y2_rad = (pi / 180) * (lat2)
    a = sin(y_dist/2)**2 + sin(x_dist/2)**2 *cos(y1_rad) * cos(y2_rad)
    c = 2 * asin(sqrt(a))
    return c * 6371 # kilometros

lon1, lat1 = -90.212452861859035, 32.316272202663704
lon2, lat2 = -88.952170968942525, 30.438559624660321

print(distancia_h(lon1, lat1, lon2, lat2))


240.6359762909508


Se puede simplificar un poco la fórmula del haverseno,  obteniendo una implementación más eficiente (en tiempo) y  con la misma precisión:

In [27]:
from math import * # importa la biblioteca o módulo de matemática

def distancia_t(lon1, lat1, lon2, lat2):
    # pre: (lon1, lat1) y (lon2, lat2) son coordenadas (lon, lat) de dos puntos en radianes.
    # post: devuelve la distancia entre los dos puntos en kilómetros
    x1, y1 = (pi / 180) * lon1,  (pi / 180) * lat1
    x2, y2 =  (pi / 180) * lon2,  (pi / 180) * lat2
    return acos(cos(y1) * cos(y2) * cos(x1 - x2) + sin(y1) * sin(y2)) * 6371

lon1, lat1 = -90.212452861859035, 32.316272202663704
lon2, lat2 = -88.952170968942525, 30.438559624660321

print(distancia_t(lon1, lat1, lon2, lat2))

240.63597629093903


**Ejemplo.** Calculemos la distancia entre las capitales de Córdoba y San Juan. En  Wikipedia podemos obtener las coordenadas decimales de cada una de esta ciudades. Hay que observar que en la Wikipedia las coordenadas geográficas son latitud, longitud y nosotros en nuestro cálculos usamos longitud, latitud. 

- Cordoba : -64.183333, -31.416667 (lon, lat),
- San Juan: -68.536389, -31.5375.

Luego la distancia entre Córdoba y San Juan es:


In [28]:
distancia_t(-64.183333, -31.416667, -68.536389, -31.5375)

413.0025113879262

Es decir 413.0025113879262 km. Podrán comprobar que podemos obtener una distancia muy parecida en Google Maps o Google Earth. 